utils/mvtree.ts

```diff
@@ -0,0 +1,62 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { AppResource } from './hostapi';
+
+/**
+ * This interface describes required properties in order to support the quick pick-through-the-tree
+ * logic. It is not required to be implemented, but the quick pick behavior will ignore the
+ * branch of the tree if these properties are not present.
+ */
+export interface QuickPickableModelNode {
+    /**
+     * An array of context values.
+     */
+    readonly contextValuesArray: string[];
+
+    /**
+     * Whether or not this node is a leaf in the overall tree
+     */
+    readonly isLeafNode: boolean;
+}
+
+/**
+ * This interface describes a provider for a subsection of the Azure Resources tree--i.e., a branch
+ * in that tree.
+ *
+ * This is largely identical to {@link vscode.TreeDataProvider}.
+ */
+export interface BranchDataProvider<T extends Partial<QuickPickableModelNode>> extends vscode.TreeDataProvider<T> {
+    /**
+     * Get the trunk element corresponding to the `resource` parameter. In essence, `resource`
+     * and the returned `T` object would occupy the same location in the tree view.
+     * @param resource The resource fetched by the Azure Resources extension.
+     * @return The model object corresponding to the resource fetched by the Azure Resources extension.
+     */
+    getTrunkElement(resource: AppResource2): vscode.ProviderResult<T>;
+
+    /**
+     * Get the children of `element`. `element` will never be falsy.
+     *
+     * This is otherwise identical to {@link vscode.TreeDataProvider.getChildren}, but redefined
+     * to reflect that `element` will always have a value.
+     * @param element The element from which the provider gets children.
+     * @return Children of `element`.
+     */
+    getChildren(element: T): vscode.ProviderResult<T[]>;

```

Comment:
It will be interesting to see how the TS compiler deals with the overlapping definitions.  If it works, great, seems like a good idea.  If it proves problematic, it's probably just as easy to note to implementors the slight changes in semantics; thorough implementors should/would be guarding against and handling the `undefined` case anyway.  In particular, for the `onDidChangeTreeData` event, we can easily treat being passed `undefined` as the equivalent to being passed the trunk element (i.e. to indicate that *everything* changed, which is still pretty much the same idea).

Reply:
IIRC it basically treats it like `override`, i.e. the nearest interface defining a method "wins". So if they say `Foo implements BranchDataProvider<Bar>`, it will treat `element` as a mandatory argument to `getChildren`, and a mandatory parameter to the event.